File: impl/lua/tests/smoke.lua
local mf = dofile("makrell_formats.lua")

local function fixture(group, file)
  return "../../shared/format-fixtures/" .. group .. "/" .. file
end

local function read_all(path)
  local file = assert(io.open(path, "r"))
  local text = file:read("*a")
  file:close()
  return text
end

local profile = mf.mbf_support_profile()
assert(profile.implemented_levels[1] == 0 and profile.implemented_levels[2] == 1, "MBF profile missing level split")

local mron = mf.parse_mron_string(read_all(fixture("mron", "sample.mron")))
assert(mron.name == "Makrell", "MRON fixture name failed")
assert(mron.stable == false, "MRON fixture bool failed")

local mrml = mf.parse_mrml_string(read_all(fixture("mrml", "sample.mrml")))
assert(mf.write_mrml_string(mrml) == '<page lang="en"><title>Makrell</title><p>A small MRML fixture.</p></page>', "MRML writer failed")

local mrtd = mf.parse_mrtd_string(read_all(fixture("mrtd", "sample.mrtd")))
assert(mrtd.records[1].name == "Ada", "MRTD fixture failed")

local conformance_mron = mf.parse_mron_string(read_all(fixture("conformance/mron", "comments-and-identifiers.mron")))
assert(conformance_mron.name == "Makrell", "Conformance MRON name failed")
assert(conformance_mron.features[2] == "typed_scalars", "Conformance MRON identifier array failed")

local block_comment_mron = mf.parse_mron_string(read_all(fixture("conformance/mron", "block-comments.mron")))
assert(block_comment_mron.name == "Makrell", "Block-comment MRON name failed")
assert(block_comment_mron.features[2] == "typed_scalars", "Block-comment MRON identifier array failed")

local negative_mron = mf.parse_mron_string(read_all(fixture("conformance/mron", "negative-numbers.mron")))
assert(negative_mron.offset == -2, "Conformance MRON negative scalar failed")
assert(negative_mron.temps[1] == -1, "Conformance MRON negative array failed")

local id_table = mf.parse_mrtd_string(read_all(fixture("conformance/mrtd", "untyped-headers.mrtd")))
assert(id_table.records[1].status == "active", "MRTD identifier values failed")
assert(id_table.columns[2].type == nil and id_table.columns[3].type == nil, "MRTD untyped headers should stay untyped")

local negative_table = mf.parse_mrtd_string(read_all(fixture("conformance/mrtd", "negative-numbers.mrtd")))
assert(negative_table.records[1].delta == -2, "Conformance MRTD negative int failed")
assert(negative_table.records[1].ratio == -3.5, "Conformance MRTD negative float failed")

local block_comment_table = mf.parse_mrtd_string(read_all(fixture("conformance/mrtd", "block-comments.mrtd")))
assert(block_comment_table.records[1].status == "active", "Block-comment MRTD first row failed")
assert(block_comment_table.records[2].note == "review", "Block-comment MRTD second row failed")

local out = mf.write_mrtd_string({
  columns = {
    { name = "name", type = "string" },
    { name = "age", type = "int" },
    { name = "active", type = "bool" },
  },
  rows = {
    { "Ada", 32, true },
    { "Ben", 41, false },
  }
})
assert(out == "name:string age:int active:bool\nAda 32 true\nBen 41 false", "MRTD writer output failed")

local untyped_out = mf.write_mrtd_string({
  columns = {
    { name = "name" },
    { name = "status" },
  },
  rows = {
    { "Ada", "active" },
  }
})
assert(untyped_out == "name status\nAda active", "MRTD untyped writer output failed")

local ok, err = pcall(function()
  mf.parse_mron_string(read_all(fixture("conformance/mron", "hyphenated-bareword.invalid.mron")))
end)
assert(not ok and err:find("Unexpected token: %-"), "Expected MRON hyphenated bareword rejection")

ok, err = pcall(function()
  mf.parse_mron_string(read_all(fixture("conformance/mron", "unclosed-array.invalid.mron")))
end)
assert(not ok and err:find("Unclosed group"), "Expected MRON unclosed array rejection")

ok, err = pcall(function()
  mf.parse_mrtd_string(read_all(fixture("conformance/mrtd", "hyphenated-bareword.invalid.mrtd")))
end)
assert(not ok and err:find("Unexpected token: %-"), "Expected MRTD hyphenated bareword rejection")

ok, err = pcall(function()
  mf.parse_mbf_level2_nodes("name value")
end)
assert(not ok and err:find("not implemented yet"), "Expected level 2 reservation error")

print("lua smoke ok")

File: impl/r/tests/smoke.R
source(file.path("R", "makrellFormats.R"))

fixture_path <- function(group, file) {
  file.path("..", "..", "shared", "format-fixtures", group, file)
}

read_fixture <- function(group, file) {
  paste(readLines(fixture_path(group, file), warn = FALSE), collapse = "\n")
}

mron <- parse_mron_file(fixture_path("mron", "sample.mron"))
stopifnot(identical(mron$name, "Makrell"))
stopifnot(identical(mron$stable, FALSE))

iddoc <- parse_mron_string("title Makrell tags [alpha beta gamma] nested { kind article status draft }")
stopifnot(identical(iddoc$title, "Makrell"))
stopifnot(identical(iddoc$tags, list("alpha", "beta", "gamma")))

conformance_mron <- parse_mron_string(read_fixture(file.path("conformance", "mron"), "comments-and-identifiers.mron"))
stopifnot(identical(conformance_mron$name, "Makrell"))
stopifnot(identical(conformance_mron$features[[2]], "typed_scalars"))

suffix_value <- apply_basic_suffix_profile("string", "2026-04-11", "dt")
stopifnot(is_basic_suffix_tagged_string(suffix_value))
stopifnot(identical(suffix_value$value, "2026-04-11"))
stopifnot(identical(suffix_value$suffix, "dt"))
stopifnot(identical(apply_basic_suffix_profile("number", "3", "k"), 3000L))
numeric_suffix <- split_basic_numeric_suffix("0.5tau")
stopifnot(identical(numeric_suffix$value, "0.5"))
stopifnot(identical(numeric_suffix$suffix, "tau"))

base_suffix_mron <- parse_mron_string(read_fixture(file.path("conformance", "mron"), "base-suffixes.mron"))
stopifnot(is_basic_suffix_tagged_string(base_suffix_mron$when))
stopifnot(identical(base_suffix_mron$when$value, "2026-04-11"))
stopifnot(identical(base_suffix_mron$bits, 10L))
stopifnot(identical(base_suffix_mron$octal, 15L))
stopifnot(identical(base_suffix_mron$mask, 255L))
stopifnot(identical(base_suffix_mron$bonus, 3000L))
stopifnot(isTRUE(all.equal(base_suffix_mron$turn, pi)))
stopifnot(isTRUE(all.equal(base_suffix_mron$angle, pi)))

block_comment_mron <- parse_mron_string(read_fixture(file.path("conformance", "mron"), "block-comments.mron"))
stopifnot(identical(block_comment_mron$name, "Makrell"))
stopifnot(identical(block_comment_mron$features[[2]], "typed_scalars"))

mrml <- parse_mrml_file(fixture_path("mrml", "sample.mrml"))
stopifnot(identical(mrml$name, "page"))
stopifnot(identical(mrml$attributes$lang, "en"))
stopifnot(identical(write_mrml_string(mrml), "<page lang=\"en\"><title>Makrell</title><p>A small MRML fixture.</p></page>"))

mrtd <- parse_mrtd_file(fixture_path("mrtd", "sample.mrtd"))
stopifnot(length(mrtd$columns) == 3)
stopifnot(identical(mrtd$records[[1]]$name, "Ada"))

idtable <- parse_mrtd_string(read_fixture(file.path("conformance", "mrtd"), "untyped-headers.mrtd"))
stopifnot(identical(idtable$records[[1]]$status, "active"))
stopifnot(identical(idtable$records[[2]]$note, "review"))
stopifnot(is.null(idtable$columns[[2]]$type))
stopifnot(is.null(idtable$columns[[3]]$type))

block_comment_table <- parse_mrtd_string(read_fixture(file.path("conformance", "mrtd"), "block-comments.mrtd"))
stopifnot(identical(block_comment_table$records[[1]]$status, "active"))
stopifnot(identical(block_comment_table$records[[2]]$note, "review"))

base_suffix_table <- parse_mrtd_string(read_fixture(file.path("conformance", "mrtd"), "base-suffixes.mrtd"))
stopifnot(is_basic_suffix_tagged_string(base_suffix_table$records[[1]]$when))
stopifnot(identical(base_suffix_table$records[[1]]$bits, 10L))
stopifnot(identical(base_suffix_table$records[[1]]$octal, 15L))
stopifnot(identical(base_suffix_table$records[[1]]$mask, 255L))
stopifnot(identical(base_suffix_table$records[[1]]$bonus, 3000L))

doc <- list(
  columns = list(
    list(name = "name", type = "string"),
    list(name = "age", type = "int"),
    list(name = "active", type = "bool")
  ),
  rows = list(
    list("Ada", 32L, TRUE),
    list("Ben", 41L, FALSE)
  )
)
stopifnot(identical(write_mrtd_string(doc), "name:string age:int active:bool\nAda 32 true\nBen 41 false"))

untyped_doc <- list(
  columns = list(
    list(name = "name", type = NULL),
    list(name = "status", type = NULL)
  ),
  rows = list(
    list("Ada", "active")
  )
)
stopifnot(identical(write_mrtd_string(untyped_doc), "name status\nAda active"))

stopifnot(inherits(try(parse_mron_string(read_fixture(file.path("conformance", "mron"), "hyphenated-bareword.invalid.mron")), silent = TRUE), "try-error"))
stopifnot(inherits(try(parse_mrtd_string(read_fixture(file.path("conformance", "mrtd"), "hyphenated-bareword.invalid.mrtd")), silent = TRUE), "try-error"))

cat("R smoke tests passed.\n")
